dopamine/discrete_domains/run_experiment.py

```diff
@@ -582,8 +590,9 @@ def __init__(
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
-    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
     self._output_lock = threading.Lock()
+    self._running_iterations = queue.Queue(max_simultaneous_iterations)
```

Comment:
It's a little harder for me to reason about concurrency when the threads are managed by limiting the size of the work queue rather than via a fixed-size thread pool.  If you want the threads to be short-lived, maybe that's the right approach.  It's just been hard for me to track when the master runs what.  I added a link to an alternative approach in another comment -- that's probably the right comment to continue this conversation in.

Reply:
Yes this is already something I have a TODO for. It also solves the environment `closing`.